### PR TITLE
[3.8] Fixed a potential use-after-free on Windows.

### PR DESCRIPTION
**Pull request by mpoeter**
### Scope & Purpose

Fix GitHub issue #15084.

Enterprise companion: https://github.com/arangodb/enterprise/pull/820

Fixed a potential use-after-free on Windows for queries that used the NeighborsEnumerator (though other PathEnumerators might have been affected as well).

The reason was that in some cases the MSVC STL unordered_map/unordered_set calculates hashes in `clear`, but the stringHeap that stored the strings which were reference by the stringRefs in the unordered_set was cleared before the set.

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] devel port: *#15266*
- [x] Backport for 3.9: *#15267*
- [x] Backport for 3.7: *#15268*

#### Related Information

- [x] Enterprise PR: https://github.com/arangodb/enterprise/pull/820
- [x] GitHub issue: #15084

### Testing & Verification

- [x] The behavior in this PR was *manually tested*

### CODE REVIEW

**Review by hkernbach — APPROVED** [review]

**Review by jsteemann — APPROVED** [review]
LGTM, but seems like some changes must be carried out for test classes derived from PathEnumerator that are used in `tests`.

File: arangod/Graph/WeightedEnumerator.cpp [hunk 1]
```diff
@@ -57,14 +57,15 @@ WeightedEnumerator::WeightedEnumerator(Traverser* traverser, TraverserOptions* o
...
-
+void WeightedEnumerator::clear() {
   _schreier.clear();
   _schreierIndex = 0;
   _lastReturned = 0;
   _queue.clear();
+}
 
+void WeightedEnumerator::setStartVertex(arangodb::velocypack::StringRef startVertex) {
+  PathEnumerator::setStartVertex(startVertex);
```
Comment (Dronplane) [inline arangod/Graph/WeightedEnumerator.cpp:68]: As I see before - `setStartVertex` implied "clear" but here I don't see `clear` call like in other enumerators. Is this intended?

Comment (jsteemann) [inline arangod/Graph/WeightedEnumerator.cpp:68]:
```suggestion
  PathEnumerator::setStartVertex(startVertex);
  clear();
```

Comment (jsteemann) [reply arangod/Graph/WeightedEnumerator.cpp:68]: Yes, it's probably better to call `clear()` here as well!

**Pull request merged**